spec/java/util/ArrayList.lsl

```diff
@@ -0,0 +1,1205 @@
+    {
+        _subListRangeCheck(fromIndex, toIndex, length);
+
+        // #problem
+        //We don't have decision about sublists.
+        //TODO
+        result = new SubList(state=Created,
+            startIndex=fromIndex,
+            endIndex=toIndex);
+    }
+
+
+    fun forEach (action: Consumer): void
+    {
+        // #problem
+        action NOT_IMPLEMENTED();
+    }
+
+
+    fun spliterator (): Spliterator
+    {
+        result = new ArrayListSpliterator(state=Initialized,
+            //This is right ? "parent=self"
+            parent=self,
+            origin = 0,
+            est=-1,
+            expectedModCount=0);
+    }
+
+
+    fun removeIf (filter: Predicate): boolean
+    {
+        // #problem
+        //We don't know how to work with lambda.
+        action NOT_IMPLEMENTED();
+    }
+
+
+    fun replaceAll (operator: UnaryOperator): void
+    {
+        // #problem
+        action NOT_IMPLEMENTED();
+    }
+
+
+    fun sort (c: Comparator): void
+    {
+        var expectedModCount = modCount;
+
+        // #problem
+        //We don't know what to do with sorting of the List.
+        action NOT_IMPLEMENTED();
+
+        if (modCount != expectedModCount)
+        {
+            action THROW_NEW("java.util.ConcurrentModificationException", []);
+        }
+
+        modCount = modCount + 1;
+    }
+}
+
+
+
+@packagePrivate
+@implements("java.util.Iterator")
+@WrapperMeta(
+    src="java.util.ArrayList$Itr",
+    //Maybe will be another name of the dst class
+    dst="ru.spbpu.libsl.overrides.collections.ArrayList_Itr",
+    forceMatchInterfaces=true)
+automaton Itr: int (
+    var cursor: int,
+    var lastRet: int = -1,
+    var expectedModCount: int
+) {
+
+    initstate Initialized;
+
+    shift Allocated -> Initialized by [
+        Itr()
+    ];
+
+    shift Initialized -> self by [
+        // read operations
+        hasNext,
+
+        // write operations
+        next,
+        remove,
+        forEachRemaining
+    ]
+
+
+    // constructors
+
+    constructor Itr()
+    {
+    }
+
+
+    // methods
+
+
+    fun hasNext (): boolean
+    {
+        result = cursor != self.parent.length;
+    }
+
+
+    fun next (): Object
+    {
+        self.parent._checkForComodification(expectedModCount);
+        var i = cursor;
+
+        if (i >= parent.length)
+        {
+            action THROW_NEW("java.util.NoSuchElementException", []);
+        }
+
+        // #problem
+        //I don't know what to do with ConcurrentModificationException(); ?
+
+        cursor = i + 1;
+        lastRet = i;
+        result = action LIST_GET(self.parent.storage, lastRet);
+    }
+
+
+    fun remove (): void
+    {
+        if (lastRet < 0)
+        {
+            action THROW_NEW("java.lang.IllegalStateException", []);
+        }
+
+        self.parent._checkForComodification(expectedModCount);
+
+        // #problem
+        //What i must to do with try-catch in this method ?
```

Comment:
This is an example of a non-deterministic behavior. Place `NOT_IMPLEMENTED` action call here.